src/object/cli_obj.c

```diff
@@ -1901,9 +1903,18 @@ obj_comp_cb(tse_task_t *task, void *data)
 		pm_stale = true;
 	}
 	if (obj_retry_error(task->dt_result)) {
-		obj_auxi->io_retry = 1;
+		/* If the RPC sponsor set the @spec_shard, then means it wants
+		 * to fetch data from the specified shard. If such shard isn't
+		 * ready for read, we should let the caller know that, but not
+		 * re-direct the fetch RPC to other replica.
+		 */
+		if (!obj_auxi->spec_shard ||
+		    (task->dt_result != -DER_STALE &&
+		     task->dt_result != -DER_INPROGRESS))
```

Comment:
why do we check these error code again?

Reply:
As the comment explained, if the RPC sponsor set the @spec_shard, then means it wants to fetch data from the specified shard. If such shard isn't ready for read, we should let the caller know that, instead of re-direct the fetch RPC to other replica. But there are some other error cases that need to retry the RPC, such as -DER_TIMEDOUT or daos_crt_network_error(). Under such case, we will retry the RPC with current shard.